kube-core/src/schema.rs

```diff
@@ -28,9 +30,530 @@ use schemars::{
 #[derive(Debug, Clone)]
 pub struct StructuralSchemaRewriter;
 
-impl Visitor for StructuralSchemaRewriter {
-    fn visit_schema_object(&mut self, schema: &mut schemars::schema::SchemaObject) {
-        schemars::visit::visit_schema_object(self, schema);
+/// A JSON Schema.
+#[allow(clippy::large_enum_variant)]
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, JsonSchema)]
+#[serde(untagged)]
+pub enum Schema {
+    /// A trivial boolean JSON Schema.
+    ///
+    /// The schema `true` matches everything (always passes validation), whereas the schema `false`
+    /// matches nothing (always fails validation).
+    Bool(bool),
+    /// A JSON Schema object.
+    Object(SchemaObject),
+}
+
+impl Schema {
+    /// Creates a new `$ref` schema.
+    ///
+    /// The given reference string should be a URI reference. This will usually be a JSON Pointer
+    /// in [URI Fragment representation](https://tools.ietf.org/html/rfc6901#section-6).
+    pub fn new_ref(reference: String) -> Self {
+        SchemaObject::new_ref(reference).into()
+    }
+
+    /// Returns `true` if `self` is a `$ref` schema.
+    ///
+    /// If `self` is a [`SchemaObject`] with `Some` [`reference`](struct.SchemaObject.html#structfield.reference) set, this returns `true`.
+    /// Otherwise, returns `false`.
+    pub fn is_ref(&self) -> bool {
+        match self {
+            Schema::Object(o) => o.is_ref(),
+            _ => false,
+        }
+    }
+
+    /// Converts the given schema (if it is a boolean schema) into an equivalent schema object.
+    ///
+    /// If the given schema is already a schema object, this has no effect.
+    ///
+    /// # Example
+    /// ```
+    /// use kube::core::schema::{Schema, SchemaObject};
+    ///
+    /// let bool_schema = Schema::Bool(true);
+    ///
+    /// assert_eq!(bool_schema.into_object(), SchemaObject::default());
+    /// ```
+    pub fn into_object(self) -> SchemaObject {
+        match self {
+            Schema::Object(o) => o,
+            Schema::Bool(true) => SchemaObject::default(),
+            Schema::Bool(false) => SchemaObject {
+                subschemas: Some(Box::new(SubschemaValidation {
+                    not: Some(Schema::Object(Default::default()).into()),
+                    ..Default::default()
+                })),
+                ..Default::default()
+            },
+        }
+    }
+}
+
+impl From<SchemaObject> for Schema {
+    fn from(o: SchemaObject) -> Self {
+        Schema::Object(o)
+    }
+}
+
+impl From<bool> for Schema {
+    fn from(b: bool) -> Self {
+        Schema::Bool(b)
+    }
+}
+
+
+/// A JSON Schema object.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Default, JsonSchema)]
+#[serde(rename_all = "camelCase", default)]
+pub struct SchemaObject {
```

Comment:
Presumably many of these types here were lifted from schemars 0.8?
Are they necessary for us or is it mostly an api convenience?

Secondly, are they necessary to stay here as `pub`? They don't seem used in the examples.